genesyscloud/task_management_workitem_schema/genesyscloud_task_management_workitem_schema_proxy.go

```diff
@@ -0,0 +1,226 @@
+	// undocumented limit or if there would be changes to the API call before release.
+
+	schemas, _, err := p.taskManagementApi.GetTaskmanagementWorkitemsSchemas()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get all workitem schemas: %v", err)
+	}
+	if schemas.Entities == nil || *schemas.Total == 0 {
+		return &([]platformclientv2.Dataschema{}), nil
+	}
+
+	return schemas.Entities, nil
+}
+
+// getTaskManagementWorkitemSchemasByNameFn is an implementation of the function to get a Genesys Cloud task management workitem schemas by name
+func getTaskManagementWorkitemSchemasByNameFn(ctx context.Context, p *taskManagementProxy, name string) (matchingSchemas *[]platformclientv2.Dataschema, retryable bool, err error) {
+	finalSchemas := []platformclientv2.Dataschema{}
+
+	schemas, err := p.getAllTaskManagementWorkitemSchema(ctx)
+	if err != nil {
+		return nil, false, err
+	}
+
+	for _, schema := range *schemas {
+		if schema.Name != nil && *schema.Name == name {
+			finalSchemas = append(finalSchemas, schema)
+		}
+	}
+
+	if len(finalSchemas) == 0 {
+		return nil, true, fmt.Errorf("no task management workitem schema found with name %s", name)
+	}
+
+	return &finalSchemas, false, nil
+}
+
+// getTaskManagementWorkitemSchemaByIdFn is an implementation of the function to get a Genesys Cloud task management workitem schema by Id
+func getTaskManagementWorkitemSchemaByIdFn(ctx context.Context, p *taskManagementProxy, id string) (schema *platformclientv2.Dataschema, statusCode int, err error) {
+	schema, resp, err := p.taskManagementApi.GetTaskmanagementWorkitemsSchema(id)
+	if err != nil {
+		return nil, resp.StatusCode, fmt.Errorf("failed to retrieve task management workitem schema by id %s: %v", id, err)
+	}
+
+	return schema, resp.StatusCode, nil
+}
+
+// updateTaskManagementWorkitemSchemaFn is an implementation of the function to update a Genesys Cloud task management workitem schema
+func updateTaskManagementWorkitemSchemaFn(ctx context.Context, p *taskManagementProxy, id string, schemaUpdate *platformclientv2.Dataschema) (*platformclientv2.Dataschema, error) {
+	schema, _, err := p.taskManagementApi.PutTaskmanagementWorkitemsSchema(id, *schemaUpdate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update task management workitem schema: %s", err)
+	}
+	return schema, nil
+}
+
+// deleteTaskManagementWorkitemSchemaFn is an implementation function for deleting a Genesys Cloud task management workitem schema
+func deleteTaskManagementWorkitemSchemaFn(ctx context.Context, p *taskManagementProxy, id string) (statusCode int, err error) {
+	resp, err := p.taskManagementApi.DeleteTaskmanagementWorkitemsSchema(id)
+	if err != nil {
+		return resp.StatusCode, fmt.Errorf("failed to delete task management workitem schema: %s", err)
+	}
+
+	return resp.StatusCode, nil
+}
+
+// getTaskManagementWorkitemSchemaDeletedStatusFn is an implementation function to get the 'deleted' status of a Genesys Cloud task management workitem schema
```

Comment:
I am confused about the method below. 

1.   Why do we need this code?  What is happening with these deleted items? 
2.  Why did we have to call this directly?  Is this not available in the SDK?

Usually, CX as Code does not care about items marked as deleted unless, like the user, the item never actually gets deleted and instead gets re-activated if it was deleted.  We also tend to care about deleted items on export.  In the getAll() function we use for export, we don't want to include items that have been deleted.  We only want the active items. (usually)

Reply:
Hi John, so what happens when a schema is deleted is that the record itself can still be queried by `GET`, it would just have a `deleted` property toggled to true (I suppose similar to user). I don't think it can be reactivated but it's different from other objects which would just give us 404 after the object deletion:

![image](https://github.com/MyPureCloud/terraform-provider-genesyscloud/assets/22146579/817d0d43-4f19-4230-9efa-1d64f8d7676f)

So the method `getTaskManagementWorkitemSchemaDeletedStatus` is to get whether the the `deleted` property is true or false. Now the reason this is called directly and not on the usual `GET` is because workitem schemas are of type [Dataschema](https://github.com/MyPureCloud/platform-client-sdk-go/blob/master/platformclientv2/dataschema.go) and `Dataschema` does not have a `deleted` property:

![image](https://github.com/MyPureCloud/terraform-provider-genesyscloud/assets/22146579/04aa5ba7-7430-4775-8371-cdc71eeabc4e)

The API response body(which has `deleted`) does not match the Dataschema struct and so the `deleted` information is lost when using the usual SDK method. I think the reason the JSON and the struct does not match is because `Dataschema` is not used exclusively for workitem schemas but is also used in some fields of external contacts and external organization. Anyway, that's the reason I created a separate method for getting deleted status. I kept the proxy `getTaskManagementWorkitemSchema` method as is using the SDK method to keep it simple.

> In the getAll() function we use for export, we don't want to include items that have been deleted. We only want the active items

The getAll API unfortunately does not have any parameters to customize the query but by default it just includes the non-`deleted` schemas so we don't have to check its results if they are deleted or not.

Reply:
Cool.  Thanks for the writeup